From: Jiri Slaby <jslaby@suse.cz>
Date: Tue, 25 Jul 2017 13:36:57 +0200
Subject: Revert "Add "shutdown" to "struct class"."
Patch-mainline: never, kabi
References: kabi

This reverts commit 5c9a2972983fca37e73648d9a3aa62a9ad048c3c, upstream
commit f77af15165847406b15d8f70c382c4cb15846b2a. It breaks kABI
heavily by changing struct class and I see no way out.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 drivers/base/core.c    | 6 +-----
 include/linux/device.h | 2 --
 2 files changed, 1 insertion(+), 7 deletions(-)

diff --git a/drivers/base/core.c b/drivers/base/core.c
index afe045792796..f18856f5954b 100644
--- a/drivers/base/core.c
+++ b/drivers/base/core.c
@@ -2094,11 +2094,7 @@ void device_shutdown(void)
 		pm_runtime_get_noresume(dev);
 		pm_runtime_barrier(dev);
 
-		if (dev->class && dev->class->shutdown) {
-			if (initcall_debug)
-				dev_info(dev, "shutdown\n");
-			dev->class->shutdown(dev);
-		} else if (dev->bus && dev->bus->shutdown) {
+		if (dev->bus && dev->bus->shutdown) {
 			if (initcall_debug)
 				dev_info(dev, "shutdown\n");
 			dev->bus->shutdown(dev);
diff --git a/include/linux/device.h b/include/linux/device.h
index 7075a2485ed3..b8f411b57dcb 100644
--- a/include/linux/device.h
+++ b/include/linux/device.h
@@ -368,7 +368,6 @@ int subsys_virtual_register(struct bus_type *subsys,
  * @suspend:	Used to put the device to sleep mode, usually to a low power
  *		state.
  * @resume:	Used to bring the device from the sleep mode.
- * @shutdown:	Called at shut-down time to quiesce the device.
  * @ns_type:	Callbacks so sysfs can detemine namespaces.
  * @namespace:	Namespace of the device belongs to this class.
  * @pm:		The default device power management operations of this class.
@@ -397,7 +396,6 @@ struct class {
 
 	int (*suspend)(struct device *dev, pm_message_t state);
 	int (*resume)(struct device *dev);
-	int (*shutdown)(struct device *dev);
 
 	const struct kobj_ns_type_operations *ns_type;
 	const void *(*namespace)(struct device *dev);
-- 
2.13.3

